agent/lib/kontena/workers/volumes/volume_manager.rb

```diff
@@ -0,0 +1,115 @@
+require_relative '../../models/volume'
+
+module Kontena::Workers::Volumes
+  class VolumeManager
+    include Celluloid
+    include Celluloid::Notifications
+    include Kontena::Logging
+    include Kontena::Helpers::RpcHelper
+    include Kontena::Helpers::WaitHelper
+    include Kontena::Observer
+
+    attr_reader :node
+
+    def initialize(autostart = true)
+      @workers = {}
+      subscribe('volume:update', :on_update_notify)
+      async.start if autostart
+    end
+
+    def start
+      observe(Actor[:node_info_worker]) do |node|
+        @node = node
+      end
+      wait_until!('waiting for node info', interval: 0.5) { self.node }
+      populate_volumes_from_docker
+      loop do
+        populate_volumes_from_master
+        sleep 30
+      end
+    end
+
+    def on_update_notify(_, _)
+      populate_volumes_from_master
+    end
+
+    def populate_volumes_from_master
+      exclusive {
+        request = rpc_client.request("/node_volumes/list", [node.id])
+        response = request.value
+        unless response['volumes'].is_a?(Array)
+          warn "failed to get list of volumes from master: #{response['error']}"
+          return
+        end
+        debug "got volumes from master: #{response}"
+
+        volumes = response['volumes']
+
+        terminate_volumes(volumes.map {|v| v.dig('labels','io.kontena.volume.id')})
+
+        volumes.each do |s|
+          ensure_volume(Kontena::Models::Volume.new(s))
+        end
+      }
+    end
+
+    def populate_volumes_from_docker
+      info "syncing volumes from docker"
+      Docker::Volume.all.each do |volume|
```

Comment:
I was thinking that this _might_ allow us to map some of the volumes to migrated grid scoped volumes. Not entirely sure yet if that's even possible/feasible.